### DISCUSSION TITLE
`tuple[Literal[1, 2]]` is not assignable to `tuple[Literal[1]] | tuple[Literal[2]]`

### DISCUSSION THREAD

**Discussion by InSyncWithFoo** [microsoft/pyright#8271] [General]
Reproducible example ([playground](https://pyright-play.net/?pyrightVersion=1.1.369&pythonVersion=3.13&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2AsAFB10AmJwUwAjABQAeAXLAFcElEgG0ipCpVHsANFABMAXSUBKKAFoAfIOFiJZKjJVQAPrpHjih6cqW86UJ1BAkYAkCijc6QA)):

```python
from typing import Literal

def f1(x: tuple[Literal[1, 2]]) -> tuple[Literal[1]] | tuple[Literal[2]]:
    return x  # error: "tuple[Literal[1, 2]]" is incompatible with "tuple[Literal[1]] | tuple[Literal[2]]"
```

`tuple[]` should distribute over the union `Literal[1, 2]`/`Literal[1] | Literal[2]`, resulting in `tuple[Literal[1]] | tuple[Literal[2]]`. Is this a bug or a known limitation?

**Comment by erictraut** [comment]
This is a known limitation. Pyright does not expand tuples when considering type compatibility. No other Python type checkers do this either.

Pyright, unlike other type checkers, does perform limited tuple expansion for purposes of type narrowing in `match` statements, but that's the only place where such expansion is performed currently.

The typing spec was recently updated to indicate that such [types are equivalent](https://typing.readthedocs.io/en/latest/spec/tuples.html#type-compatibility-rules). It indicates that type checkers _may_ treat them as such for purposes of narrowing tuple types, but it doesn't mandate it.
